l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -0,0 +1,223 @@
+    }
+
+    /* solhint-disable no-unused-vars */
+    function bridgeMint(uint256 _chainId, bytes32 _assetInfo, bytes calldata _data) external payable override {
+        // if (!hyperbridgingEnabled[_chainId]) { // ToDo: add logic & uncomment
+        //     // Add back
+        //     // Check that the chain has sufficient balance
+        //     require(chainBalance[_chainId][l1Token] >= _amount, "NTV not enough funds 2"); // not enough funds
+        //     chainBalance[_chainId][l1Token] -= _amount;
+        // }
+        address l1Token = tokenAddress[_assetInfo];
+        (uint256 _amount, address l1Receiver) = abi.decode(_data, (uint256, address));
+        if (l1Token == ETH_TOKEN_ADDRESS) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), l1Receiver, _amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "NTV: withdraw failed");
+        } else {
+            // Withdraw funds
+            IERC20(l1Token).safeTransfer(l1Receiver, _amount);
+        }
+    }
+
+    function bridgeClaimFailedBurn(
+        uint256 _chainId,
+        bytes32 _assetInfo,
+        address,
+        bytes calldata _data
+    ) external payable override {
+        (uint256 _amount, address _depositSender) = abi.decode(_data, (uint256, address));
+        address l1Token = tokenAddress[_assetInfo];
+        require(_amount > 0, "y1");
+
+        if (l1Token == ETH_TOKEN_ADDRESS) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), _depositSender, _amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "NTV: claimFailedDeposit failed");
+        } else {
+            IERC20(l1Token).safeTransfer(_depositSender, _amount);
+            // Note we don't allow weth deposits anymore, but there might be legacy weth deposits.
+            // until we add Weth bridging capabilities, we don't wrap/unwrap weth to ether.
+        }
+
+        if (L1_SHARED_BRIDGE.hyperbridgingEnabled(_chainId)) {
+            // check that the chain has sufficient balance
+            require(chainBalance[_chainId][l1Token] >= _amount, "NTV n funds");
+            chainBalance[_chainId][l1Token] -= _amount;
+        }
+    }
+
+    function getAssetInfoFromLegacy(address _l1TokenAddress) public view override returns (bytes32) {
+        if (tokenAddress[getAssetInfo(_l1TokenAddress)] != address(0)) {
+            return getAssetInfo(_l1TokenAddress);
+        }
+        return bytes32(uint256(uint160(_l1TokenAddress)));
+    }
+
+    function getAssetInfo(address _l1TokenAddress) public view override returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    block.chainid,
+                    NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS,
```

Comment:
I think the purpose is to easily predict the assetInfo when finalizing withdrawals / deposits from other chains

Reply:
We discussed with Stas, what we could actually do is have the L1SharedBridge send the chosen L2AssetHAndler with the txs. This would work in the general case. 

Alternatively the L1NativeTokenVault could send an L1->L2 message to the L2SharedBridge to register the token. 

This second one might be a bit more fragile, but probably still ok.